libraries/freertos_plus/standard/https/src/iot_https_client.c

```diff
@@ -0,0 +1,1829 @@
+ * sends body in the response, but it exceeds the size of  _httpResponse->pHeaders buffer, then we need to flush the 
+ * network buffer.
+ * 
+ * If the amount of body received on the network does not fit into a non-null IotHttpsResponseInfo_t.syncInfo.respData, 
+ * then we need to flush the network buffer.
+ * 
+ * If an asynchronous request cancels in the middle of a response process, after already sending the request message, 
+ * then we need to flush the network buffer.
+ * 
+ * @param[in] _httpsConnection - HTTP connection context.
+ * @param[in] _httpsResponse - HTTP response context.
+ * 
+ * @return #IOT_HTTPS_OK if we received successfully flushed the network data.
+ *         #IOT_HTTPS_PARSING_ERROR if there was an error with parsing the data.
+ *         #IOT_HTTPS_NETWORK_ERROR if there was an error receiving the data on the network.
+ *         #IOT_HTTPS_TIMEOUT_ERROR if we timedout trying to receive data from the network.
+ */
+static IotHttpsReturnCode_t _flushHttpsNetworkData( _httpsConnection_t* _httpsConnection, _httpsResponse_t* _httpsResponse );
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief A map of the method enum to strings 
+ * 
+ * These are in order to the HTTP request method enums defined in IotHttpsMethod_t.
+ */
+static const char* httpsMethodStrings[] = {
+    "GET",
+    "HEAD"
+};
+
+/**
+ * @brief Definition of the http-parser settings.
+ * 
+ * The http_parser_settings holds all of the callbacks invoked by the http-parser.
+ */
+static http_parser_settings _httpParserSettings = { 0 };
+
+/*-----------------------------------------------------------*/
+
+static int _httpParserOnMessageBeginCallback(http_parser * pHttpParser)
+{
+    IotLogDebug("Parser: Start of HTTPS Response message.");
+    _httpsResponse_t *_httpsResponse = (_httpsResponse_t*)(pHttpParser->data);
+    /* Set the state of the parser. The headers are at the start of the message always. */
+    _httpsResponse->parserState = PARSER_STATE_IN_HEADERS;
+    return 0;
+}
+
+/*-----------------------------------------------------------*/
+
+static int _httpParserOnStatusCallback(http_parser * pHttpParser, const char * pLoc, size_t length)
+{
+    _httpsResponse_t *_httpsResponse = (_httpsResponse_t*)(pHttpParser->data);
+    IotLogDebug("Parser: Status %.*s retrieved from HTTPS response.", length, pLoc);
+    /* Save the status code so it can be retrieved with IotHttpsClient_ReadResponseStatus(). */
+    _httpsResponse->status = pHttpParser->status_code;
+    /* If we are parsing the network data received in the header buffer then we can increment 
+      _httpsResponse->pHeadersCur. The status line in the response is part of the data stored in
+      _httpResponse->pHeaders. */
+    if(_httpsResponse->bufferProcessingState == PROCESSING_STATE_FILLING_HEADER_BUFFER)
+    {
+        _httpsResponse->pHeadersCur = (uint8_t*)(pLoc += length);
```

Comment:
Do we need any bound checks to make sure ```pHeadersCur``` doesnt point to outside its allowed bounds ?

Reply:
That will never happen because the parser is bounded when we first execute:
```
static IotHttpsReturnCode_t _parseHttpsMessage(http_parser* pHttpParser, char* pBuf, size_t len)
{
    size_t parsedBytes = http_parser_execute(pHttpParser, &_httpParserSettings, pBuf, len );
```

Reply:
ok